### PR TITLE
Failover doc

### PR DESCRIPTION
**Pull request by madolson**
https://github.com/redis/redis/pull/8315

### CODE REVIEW

File: commands/failover.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+This command will start a coordinated failover between a master and one of its replicas. The failover is not synchronous, instead
+a background task will handle coordinating the failover. It is designed to limit 
+data loss and unavailability of the cluster during the failover. This command is analogous
```
Comment (oranagra) [inline commands/failover.md:3]: i think we should avoid manual word wrapping, and start every sentence in a new line.
see my recent changes here (aimed to create diffs that are easier to review)
https://github.com/redis/redis-doc/pull/1485

File: commands/failover.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+If the previous master had additional replicas attached to it, they will continue replicating from it as chained replicas. You will need to manually
+execute a `REPLICAOF` on these replicas to start replicating directly from the new master.
+
+## Optional arguments
+
+The following optional arguments exist to modify the behavior of the failover flow:
+* `TIMEOUT` *milliseconds* -- This option allows specifying a maximum time a master will wait in the `waiting-for-sync` state before aborting the failover attempt and rolling back. This is intended to set an upper bound on the write outage the redis cluster can experience. Failovers typically happen in less than a second, but could take longer if there is a large amount of write traffic or the replica
+is already behind in consuming the replication stream. If this value is not specified, the timeout can be considered to be "infinite".
+* `TO` *HOST* *IP* -- This option allows designating a specific replica, by it's host and port, to failover to. The master will wait specifically for this replica to catch up to it's replication offset, and then failover to it.
```
Comment (oranagra) [inline commands/failover.md:25]:
```suggestion
* `TO` *HOST* *PORT* -- This option allows designating a specific replica, by it's host and port, to failover to. The master will wait specifically for this replica to catch up to it's replication offset, and then failover to it.
```

File: commands/failover.md (hunk 1 above)
Comment (madolson) [reply commands/failover.md:3]: Ack, good to know about

**Comment by madolson** [comment]
No worries on being petty, I didn't understand your previous comment, but I think I get it now.

File: commands/failover.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+If the previous master had additional replicas attached to it, they will continue replicating from it as chained replicas. You will need to manually execute a `REPLICAOF` on these replicas to start replicating directly from the new master.
+
+## Optional arguments
+The following optional arguments exist to modify the behavior of the failover flow:
+
+`TIMEOUT` *milliseconds* -- This option allows specifying a maximum time a master will wait in the `waiting-for-sync` state before aborting the failover attempt and rolling back.
+This is intended to set an upper bound on the write outage the redis cluster can experience.
+Failovers typically happen in less than a second, but could take longer if there is a large amount of write traffic or the replica is already behind in consuming the replication stream. 
+If this value is not specified, the timeout can be considered to be "infinite".
```
Comment (itamarhaber) [inline commands/failover.md:25]:
```suggestion
* `TIMEOUT` *milliseconds* -- This option allows specifying a maximum time a master will wait in the `waiting-for-sync` state before aborting the failover attempt and rolling back.
    This is intended to set an upper bound on the write outage the Redis cluster can experience.
    Failovers typically happen in less than a second, but could take longer if there is a large amount of write traffic or the replica is already behind in consuming the replication stream. 
    If this value is not specified, the timeout can be considered to be "infinite".
```

File: commands.json [hunk 4]
```diff
@@ -1223,6 +1223,45 @@
...
+          },
+          {
+            "command": "FORCE",
+            "optional": true
+          }
+        ]
+      },
+      {
+        "command": "ABORT",
+        "optional": true
```
[the remark this answers is not in the document]
Comment (madolson) [reply commands.json:1253]: It is optional, fixed in the PR

File: commands/failover.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+The failover is not synchronous, instead a background task will handle coordinating the failover. 
+It is designed to limit data loss and unavailability of the cluster during the failover.
+This command is analogous to the `CLUSTER FAILOVER` command for non-clustered Redis and is similar to the failover support provided by sentinel.
+The specific details of the default failover flow are as follows:
+1. The master will internally start a `CLIENT PAUSE WRITE`, which will pause incoming writes and prevent the accumulation of new data in the replication stream.
+2. The master will monitor its replicas, waiting for a replica to indicate that it has fully consumed the replication stream. If the master has multiple replicas, it will only wait for the first replica to catch up.
+3. The master will then demote itself to a replica. This is done to prevent any dual master scenarios. NOTE: The master will not discard it's data, so it will be able to rollback if the replica rejects the failover request in the next step.
+4. The previous master will send a special PSYNC request to the target replica, `PSYNC FAILOVER `, instructing the target replica to become a master.
```
Comment (oranagra) [inline commands/failover.md:9]:
```suggestion
4. The previous master will send a special PSYNC request to the target replica, `PSYNC FAILOVER`, instructing the target replica to become a master.
```

File: commands/failover.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+This command will start a coordinated failover between a master and one of its replicas.
```
[the remark this answers is not in the document]
Comment (madolson) [reply commands/failover.md:1]: It's wordy, but okay.

File: commands/failover.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+The failover is not synchronous, instead a background task will handle coordinating the failover. 
+It is designed to limit data loss and unavailability of the cluster during the failover.
+This command is analogous to the `CLUSTER FAILOVER` command for non-clustered Redis and is similar to the failover support provided by sentinel.
+The specific details of the default failover flow are as follows:
```
[the remark this answers is not in the document]
Comment (madolson) [reply commands/failover.md:5]: Sure

File: commands/failover.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+1. The master will internally start a `CLIENT PAUSE WRITE`, which will pause incoming writes and prevent the accumulation of new data in the replication stream.
+2. The master will monitor its replicas, waiting for a replica to indicate that it has fully consumed the replication stream. If the master has multiple replicas, it will only wait for the first replica to catch up.
+3. The master will then demote itself to a replica. This is done to prevent any dual master scenarios. NOTE: The master will not discard it's data, so it will be able to rollback if the replica rejects the failover request in the next step.
+4. The previous master will send a special PSYNC request to the target replica, `PSYNC FAILOVER `, instructing the target replica to become a master.
+5. Once the previous master receives acknowledgement the `PSYNC FAILOVER` was accepted it will unpause its clients. If the PSYNC request is rejected, the master will abort the failover and return to normal.
+
+The field `master_failover_state` in `INFO REPLICATION` can be used to track the current state of the failover, which has the following values:
```
[the remark this answers is not in the document]
Comment (madolson) [reply commands/failover.md:12]: Can do, what's the motivation here for the second argument being lower case?

File: wordlist [hunk 9]
```diff
@@ -152,6 +152,7 @@ Yukihiro
 ZPOP
 ZSET
 addr
+acknowledgement
```
[the remark this answers is not in the document]
Comment (madolson) [reply wordlist:155]: wut

File: commands.json [hunk 10]
```diff
@@ -1223,6 +1223,45 @@
     "since": "1.2.0",
     "group": "generic"
   },
+  "FAILOVER": {
+    "summary": "Start a coordinated failover between this server and one of its replicas.",
```
[the remark this answers is not in the document]
Comment (madolson) [reply commands.json:1227]: I tried to write the whole thing PC, but using anything but master is going to be confusing.

File: commands/failover.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+`TIMEOUT` *milliseconds* -- This option allows specifying a maximum time a master will wait in the `waiting-for-sync` state before aborting the failover attempt and rolling back.
+This is intended to set an upper bound on the write outage the redis cluster can experience.
+Failovers typically happen in less than a second, but could take longer if there is a large amount of write traffic or the replica is already behind in consuming the replication stream. 
+If this value is not specified, the timeout can be considered to be "infinite".
+
+`TO` *HOST* *PORT* -- This option allows designating a specific replica, by its host and port, to failover to. The master will wait specifically for this replica to catch up to it's replication offset, and then failover to it.
+
+`FORCE` -- If both the `TIMEOUT` and `TO` options are set, the force flag can also be used to designate that that once the timeout has elapsed, the master should failover to the target replica any way.
+This can be used for a best effort attempt at a failover without data loss, but limiting write outage.
```
[the remark this answers is not in the document]
Comment (madolson) [reply commands/failover.md:30]: I think write outage is correct.

File: commands/failover.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (itamarhaber) [reply commands/failover.md]: Historically, INFO's sections are not considered "subcommands". Subcommands are upper case. But in any/either case, consistency is futile.

File: commands/failover.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (itamarhaber) [reply commands/failover.md]: Sure, sorry, amending.

File: wordlist (hunk 9 above)
[the remark this answers is not in the document]
Comment (itamarhaber) [reply wordlist:155]: Potato, potato.

File: commands/failover.md [hunk 12]
```diff
@@ -19,21 +20,22 @@ If the previous master had additional replicas attached to it, they will continu
 ## Optional arguments
 The following optional arguments exist to modify the behavior of the failover flow:
 
-`TIMEOUT` *milliseconds* -- This option allows specifying a maximum time a master will wait in the `waiting-for-sync` state before aborting the failover attempt and rolling back.
-This is intended to set an upper bound on the write outage the redis cluster can experience.
-Failovers typically happen in less than a second, but could take longer if there is a large amount of write traffic or the replica is already behind in consuming the replication stream. 
-If this value is not specified, the timeout can be considered to be "infinite".
+* `TIMEOUT` *milliseconds* -- This option allows specifying a maximum time a master will wait in the `waiting-for-sync` state before aborting the failover attempt and rolling back.
+    This is intended to set an upper bound on the write outage the Redis cluster can experience.
```
Comment (oranagra) [inline commands/failover.md:24]: What is this indentation for?  AFAIK it's not needed

**Review by oranagra — COMMENTED** [review]
LGTM (although it's very hard to see what was changed, and I'm too lazy to read it all again)

File: commands/failover.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (madolson) [reply commands/failover.md]: Sounds good, I try to conform to consistency :)

File: commands/failover.md (hunk 12 above)
Comment (madolson) [reply commands/failover.md:24]: It was in Itamar's suggested edit's, I presume for readability? I was also under the impression just stripped out leading spaces, but, I don't know if we do anything fancy during rendering.

I didn't change any content, just typos and formatting.

Comment (itamarhaber) [reply commands/failover.md:24]: Not needed indeed, my fat finger on the tab or something

Comment (itamarhaber) [reply commands/failover.md:24]: The indention is needed so these lines are a part of the bullet.

Comment (itamarhaber) [reply commands/failover.md:24]: Sorry, not needed. Pushing a fix for that and also for a missing newline above.

Comment (madolson) [reply commands/failover.md]: Ack.

**Pull request merged**